politeiawww/user/cockroachdb/cockroachdb_test.go

```diff
@@ -0,0 +1,999 @@
+	"github.com/google/uuid"
+	"github.com/jinzhu/gorm"
+	_ "github.com/jinzhu/gorm/dialects/postgres"
+)
+
+// Custom go-sqlmock types for type assertion
+type AnyBlob struct{}
+type AnyTime struct{}
+
+func (a AnyBlob) Match(v driver.Value) bool {
+	_, ok := v.([]byte)
+	return ok
+}
+
+func (a AnyTime) Match(v driver.Value) bool {
+	_, ok := v.(time.Time)
+	return ok
+}
+
+// Helpers
+func newPaywallAddressIndex(t *testing.T, i uint64) *[]byte {
+	t.Helper()
+
+	index := make([]byte, 8)
+	binary.LittleEndian.PutUint64(index, i)
+	return &index
+}
+
+func newCdbUser(t *testing.T, cdb *cockroachdb) (User, []byte) {
+	t.Helper()
+
+	uuid := uuid.New()
+	u := user.User{
+		ID:       uuid,
+		Username: "test" + uuid.String(),
+	}
+
+	// Make user identity
+	fid, err := identity.New()
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+	id, err := user.NewIdentity(hex.EncodeToString(fid.Public.Key[:]))
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+	u.Identities = append(u.Identities, *id)
+
+	// Make user blob
+	eu, err := user.EncodeUser(u)
+	if err != nil {
+		t.Fatalf("%s", err)
+	}
+	eb, err := cdb.encrypt(user.VersionUser, eu)
+	if err != nil {
+		t.Fatalf("%s", err)
+	}
+
+	return convertUserFromUser(u, eb), eb
+}
+
+func setupTestDB(t *testing.T) (*cockroachdb, sqlmock.Sqlmock, func()) {
+	t.Helper()
+
+	db, mock, err := sqlmock.New()
+	if err != nil {
+		t.Fatalf("error %s while creating stub db conn", err)
+	}
+
+	gdb, err := gorm.Open("postgres", db)
+	if err != nil {
+		t.Fatalf("error %s while opening db with gorm", err)
+	}
+
+	b := []byte("random")
+	var key [32]byte
+	copy(key[:], b)
+
+	c := &cockroachdb{
+		userDB:        gdb,
+		encryptionKey: &key,
+	}
+
+	return c, mock, func() {
+		db.Close()
+	}
+}
+
+// Tests
+func TestSetPaywallAddressIndex(t *testing.T) {
+	cdb, mock, close := setupTestDB(t)
+	defer close()
+
+	// Arguments
+	i := uint64(1)
+	index := newPaywallAddressIndex(t, i)
+
+	// Query
+	sql := `UPDATE "key_value" SET "value" = $1 WHERE "key_value"."key" = $2`
+
+	// Expectations
+	mock.ExpectBegin()
+	mock.ExpectExec(regexp.QuoteMeta(sql)).
+		WithArgs(index, keyPaywallAddressIndex).
+		WillReturnResult(sqlmock.NewResult(0, 1))
+	mock.ExpectCommit()
+
+	// Execute method to match queries
+	err := cdb.SetPaywallAddressIndex(i)
+	if err != nil {
+		t.Errorf("error creating new user %s", err)
+	}
+
+	// Make sure query expectations were met
+	err = mock.ExpectationsWereMet()
+	if err != nil {
+		t.Errorf("unfulfilled expectations: %s", err)
+	}
+}
+
+func TestUserNewSuccess(t *testing.T) {
+	cdb, mock, close := setupTestDB(t)
+	defer close()
+
+	// Arguments
+	index := newPaywallAddressIndex(t, 1)
+	usr := user.User{
+		Email:    "test@test.com",
+		Username: "test",
+	}
+
+	// Queries
+	sqlSelectIndex := `SELECT * FROM "key_value" WHERE "key_value"."key" = $1`
+
+	sqlInsertUser := `INSERT INTO "users" ("id","username","blob","created_at","updated_at") VALUES ($1,$2,$3,$4,$5) RETURNING "users"."id"`
```

Comment:
Use multi line string literals to prevent exceeding 80 column

Reply:
I did try to use multiline strings within the queries, but it seems that `go-sqlmock` interprets the line break as part of the query, and the test can't match the executed query from our db method to the escaped sql query in the tests, for some reason.

Reply:
You should be able to do something like this.

```
	sqlInsertUser := `INSERT INTO "users" ` +
		`("id","username","blob","created_at","updated_at") ` +
		`VALUES ($1,$2,$3,$4,$5) ` +
		`RETURNING "users"."id"`
```

Reply:
did the trick :) thanks